docs/spec_draft.md

```diff
@@ -0,0 +1,124 @@
+  * **Boolean types** referred to in the document as `pred`
+  * **Floating point types**
+    * Single precision `f32`, double precision `f64` and half precision `f16`
+    floating points complying with [IEEE 754
+    format](https://ieeexplore.ieee.org/document/8766229).
+    * Bfloat16 `bf16` floating point complying with [Brain Floating Point Format](https://cloud.google.com/blog/products/ai-machine-learning/bfloat16-the-secret-to-high-performance-on-cloud-tpus).
+    Provides the same number of exponent bits as `f32`, so that it matches its
+    dynamic range, but with greatly reduced precision.
+ * **Complex types** represents a pair of floating point types. Supported ones
+ are `c64` (represents paired `f32`) and `c128` (represents paired `f64`).
+
+StableHLO supports a Tensor type `tensor`, to model the type of a n-dimensional
+array, represented in IR as `tensor<SxE>` such that
+
+  * Shape `S` is a list of number of elements in each of the
+  dimensions and represented, in increasing order of the corresponding dimension
+  number, as an array of values of type `ui64`. A 0-value represents missing or
+  empty data in the corresponding dimension.
+  * Element type `E` is any one of the supported element types mentioned above.
+
+## Programs
+
+StableHLO programs consist of functions. Each function has arguments and results
+of supported types and a list of ops in static single-assignment (SSA) form
+which is terminated by a return op which produces the results of the function.
+StableHLO ops take operands and produce results.
+
+```
+func.func @example_func(%arg: tensor<4x16xf32>) -> tensor<4x16xf32>  {
+ %1 = stablehlo.floor %arg : tensor<4x16xf32>
+ %2 = stablehlo.ceil  %arg : tensor<4x16xf32>
+ %3 = stablehlo.add %1, %2 : tensor<4x16xf32>
+ func.return %3 : tensor<4x16xf32>
+}
+```
+
+A program is executed by passing argument values to a given function and
+computing result values. Result values of a function are computed by evaluating
+the graph of ops rooted in the corresponding return op. The evaluation order is
+implementation-defined, as long as ops are evaluated before their uses. Possible
+execution orders of the above example program are `%1` → `%2` → `%3` → `return`
+or `%2` → `%1` → `%3` → `return`.
+
+## Constants
+
+The section describes the constants supported in StableHLO along with their
+syntax.
+
+  * **Integer Constants** Standard integers are constants of the integer type
+  (signed or unsigned). Negative numbers can be used with signed integer types.
+  * **Boolean Constants** "true" and "false" are both valid constants of the
+  `pred` type.
+  * **Floating point Constants** Floating point constants use standard decimal
+  notation, e.g., "123.421", exponential notation, e.g. "1.23421e+2", or a more
+  precise hexadecimal notation, e.g., "0x42f6d78d".
+  * **Complex Constants** Complex constants are represented as a pair of real
+  and imaginary values of `f32` or `f64` types .
```

Comment:
Let's define a notation for these constants, e.g. "E.g. `(12.34, 56.78)`" or whatever syntax you prefer - I don't have a strong preference here.

Reply:
done.